sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala

```diff
@@ -0,0 +1,153 @@
+  private var eof = false
+
+  override def hasNext: Boolean = {
+    maybeReadNext()
+    byteArrayPair != null
+  }
+
+  override def next(): ByteArrayPair = {
+    maybeReadNext()
+    val nextByteArrayPair = byteArrayPair
+    byteArrayPair = null
+    nextByteArrayPair
+  }
+
+  def close(): Unit = { if (input != null) input.close() }
+
+  private def maybeReadNext(): Unit = {
+    if (!eof && byteArrayPair == null) {
+      val keySize = input.readInt()
+      if (keySize == -1) {
+        eof = true
+      } else if (keySize < 0) {
+        throw new IOException(
+          s"Error reading streaming state file $fileToRead of $this: key size cannot be $keySize")
+      } else {
+        val keyBuffer = new Array[Byte](keySize)
+        ByteStreams.readFully(input, keyBuffer, 0, keySize)
+        val valueSize = input.readInt()
+        if (valueSize < 0) {
+          byteArrayPair = new ByteArrayPair(keyBuffer, null)
+        } else {
+          val valueBuffer = new Array[Byte](valueSize)
+          ByteStreams.readFully(input, valueBuffer, 0, valueSize)
+          byteArrayPair = new ByteArrayPair(keyBuffer, valueBuffer)
+        }
+      }
+    }
+  }
+}
+
+class ChangelogWriter(fm: CheckpointFileManager, file: Path,
+                      compressionCodec: CompressionCodec) extends Logging {
+  private def compressStream(outputStream: DataOutputStream): DataOutputStream = {
+    val compressed = compressionCodec.compressedOutputStream(outputStream)
+    new DataOutputStream(compressed)
+  }
+
+  private var backingFileStream: CancellableFSDataOutputStream =
+    fm.createAtomic(file, overwriteIfPossible = true)
+  private var compressedStream: DataOutputStream = compressStream(backingFileStream)
+
+  def put(key: Array[Byte], value: Array[Byte]): Unit = {
+    assert(compressedStream != null)
+    compressedStream.writeInt(key.size)
+    compressedStream.write(key)
+    compressedStream.writeInt(value.size)
+    compressedStream.write(value)
+  }
+
+  def delete(key: Array[Byte]): Unit = {
+    assert(compressedStream != null)
+    compressedStream.writeInt(key.size)
+    compressedStream.write(key)
+    compressedStream.writeInt(-1)
+  }
+
+  def abort(): Unit = {
+    try {
+      if (backingFileStream != null) backingFileStream.cancel()
+      if (compressedStream != null) IOUtils.closeQuietly(compressedStream)
+    } catch {
+      // Closing the compressedStream causes the stream to write/flush flush data into the
+      // rawStream. Since the rawStream is already closed, there may be errors.
+      // Usually its an IOException. However, Hadoop's RawLocalFileSystem wraps
+      // IOException into FSError.
+      case e: FSError if e.getCause.isInstanceOf[IOException] =>
+
+      // SPARK-42668 - Catch and log any other exception thrown while trying to cancel
+      // raw stream or close compressed stream.
+      case NonFatal(ex) =>
+        logInfo(s"Failed to cancel delta file for provider " +
```

Comment:
Can we also log the delta file name/path ?

Reply:
Sure.